In [1]:
using ITensors

# Compare RDM Approaches

In [4]:
s = siteinds("S=1/2", 6);
ψ = randomMPS(s; linkdims=3)

MPS
[1] ((dim=2|id=113|"S=1/2,Site,n=1"), (dim=3|id=423|"Link,l=1"))
[2] ((dim=3|id=423|"Link,l=1"), (dim=2|id=188|"S=1/2,Site,n=2"), (dim=3|id=992|"Link,l=2"))
[3] ((dim=3|id=992|"Link,l=2"), (dim=2|id=520|"S=1/2,Site,n=3"), (dim=3|id=41|"Link,l=3"))
[4] ((dim=3|id=41|"Link,l=3"), (dim=2|id=649|"S=1/2,Site,n=4"), (dim=3|id=261|"Link,l=4"))
[5] ((dim=3|id=261|"Link,l=4"), (dim=2|id=850|"S=1/2,Site,n=5"), (dim=2|id=776|"Link,l=5"))
[6] ((dim=2|id=776|"Link,l=5"), (dim=2|id=751|"S=1/2,Site,n=6"))


Choose site 3

In [8]:
ψ1 = deepcopy(ψ);

In [12]:
orthogonalize!(ψ1, 3);

In [17]:
ρ = prime(ψ1[3], s[3]) * dag(ψ1[3]);

In [18]:
@show ρ

ρ = ITensor ord=2
Dim 1: (dim=2|id=520|"S=1/2,Site,n=3")'
Dim 2: (dim=2|id=520|"S=1/2,Site,n=3")
NDTensors.Dense{Float64, Vector{Float64}}
 2×2
  0.7864668893819489    -0.020836894595178346
 -0.020836894595178346   0.2135331106180514


ITensor ord=2 (dim=2|id=520|"S=1/2,Site,n=3")' (dim=2|id=520|"S=1/2,Site,n=3")
NDTensors.Dense{Float64, Vector{Float64}}

Check trace

In [19]:
tr(ρ)

1.0000000000000002

check hermitian

In [23]:
isequal(ρ, dag(ρ))

true

# Approach 2 Manually tracing

In [47]:
ψ2 = deepcopy(ψ)

MPS
[1] ((dim=2|id=113|"S=1/2,Site,n=1"), (dim=3|id=423|"Link,l=1"))
[2] ((dim=3|id=423|"Link,l=1"), (dim=2|id=188|"S=1/2,Site,n=2"), (dim=3|id=992|"Link,l=2"))
[3] ((dim=3|id=992|"Link,l=2"), (dim=2|id=520|"S=1/2,Site,n=3"), (dim=3|id=41|"Link,l=3"))
[4] ((dim=3|id=41|"Link,l=3"), (dim=2|id=649|"S=1/2,Site,n=4"), (dim=3|id=261|"Link,l=4"))
[5] ((dim=3|id=261|"Link,l=4"), (dim=2|id=850|"S=1/2,Site,n=5"), (dim=2|id=776|"Link,l=5"))
[6] ((dim=2|id=776|"Link,l=5"), (dim=2|id=751|"S=1/2,Site,n=6"))


In [56]:
ψ2_dag = prime(linkinds, dag(ψ2))

MPS
[1] ((dim=2|id=113|"S=1/2,Site,n=1"), (dim=3|id=423|"Link,l=1")')
[2] ((dim=3|id=423|"Link,l=1")', (dim=2|id=188|"S=1/2,Site,n=2"), (dim=3|id=992|"Link,l=2")')
[3] ((dim=3|id=992|"Link,l=2")', (dim=2|id=520|"S=1/2,Site,n=3"), (dim=3|id=41|"Link,l=3")')
[4] ((dim=3|id=41|"Link,l=3")', (dim=2|id=649|"S=1/2,Site,n=4"), (dim=3|id=261|"Link,l=4")')
[5] ((dim=3|id=261|"Link,l=4")', (dim=2|id=850|"S=1/2,Site,n=5"), (dim=2|id=776|"Link,l=5")')
[6] ((dim=2|id=776|"Link,l=5")', (dim=2|id=751|"S=1/2,Site,n=6"))


In [57]:
L = ψ2[1] * ψ2_dag[1] * ψ2[2] * ψ2_dag[2]

ITensor ord=2 (dim=3|id=992|"Link,l=2") (dim=3|id=992|"Link,l=2")'
NDTensors.Dense{Float64, Vector{Float64}}

In [58]:
R = ψ2[4] * ψ2_dag[4] * ψ2[5] * ψ2_dag[5] * ψ2[6] * ψ2_dag[6]

ITensor ord=2 (dim=3|id=41|"Link,l=3") (dim=3|id=41|"Link,l=3")'
NDTensors.Dense{Float64, Vector{Float64}}

In [65]:
ψ2[3]

ITensor ord=3 (dim=3|id=992|"Link,l=2") (dim=2|id=520|"S=1/2,Site,n=3") (dim=3|id=41|"Link,l=3")
NDTensors.Dense{Float64, Vector{Float64}}

In [70]:
out = ψ2[3] * ψ2[3]' * L * R

ITensor ord=2 (dim=2|id=520|"S=1/2,Site,n=3") (dim=2|id=520|"S=1/2,Site,n=3")'
NDTensors.Dense{Float64, Vector{Float64}}

In [72]:
@show out

out = ITensor ord=2
Dim 1: (dim=2|id=520|"S=1/2,Site,n=3")
Dim 2: (dim=2|id=520|"S=1/2,Site,n=3")'
NDTensors.Dense{Float64, Vector{Float64}}
 2×2
  0.7864668893819488   -0.02083689459517837
 -0.02083689459517837   0.21353311061805133


ITensor ord=2 (dim=2|id=520|"S=1/2,Site,n=3") (dim=2|id=520|"S=1/2,Site,n=3")'
NDTensors.Dense{Float64, Vector{Float64}}

Do the same but for site 6

In [73]:
ψ3 = deepcopy(ψ)

MPS
[1] ((dim=2|id=113|"S=1/2,Site,n=1"), (dim=3|id=423|"Link,l=1"))
[2] ((dim=3|id=423|"Link,l=1"), (dim=2|id=188|"S=1/2,Site,n=2"), (dim=3|id=992|"Link,l=2"))
[3] ((dim=3|id=992|"Link,l=2"), (dim=2|id=520|"S=1/2,Site,n=3"), (dim=3|id=41|"Link,l=3"))
[4] ((dim=3|id=41|"Link,l=3"), (dim=2|id=649|"S=1/2,Site,n=4"), (dim=3|id=261|"Link,l=4"))
[5] ((dim=3|id=261|"Link,l=4"), (dim=2|id=850|"S=1/2,Site,n=5"), (dim=2|id=776|"Link,l=5"))
[6] ((dim=2|id=776|"Link,l=5"), (dim=2|id=751|"S=1/2,Site,n=6"))


In [74]:
orthogonalize!(ψ3, 6);

In [80]:
ρ = prime(ψ3[6], s[6]) * dag(ψ3[6]);
@show ρ

ρ = ITensor ord=2
Dim 1: (dim=2|id=751|"S=1/2,Site,n=6")'
Dim 2: (dim=2|id=751|"S=1/2,Site,n=6")
NDTensors.Dense{Float64, Vector{Float64}}
 2×2
  0.5871178549425848   -0.09668333287052744
 -0.09668333287052744   0.41288214505741566


ITensor ord=2 (dim=2|id=751|"S=1/2,Site,n=6")' (dim=2|id=751|"S=1/2,Site,n=6")
NDTensors.Dense{Float64, Vector{Float64}}

In [81]:
tr(ρ)

1.0000000000000004

In [82]:
ψ4 = deepcopy(ψ)

MPS
[1] ((dim=2|id=113|"S=1/2,Site,n=1"), (dim=3|id=423|"Link,l=1"))
[2] ((dim=3|id=423|"Link,l=1"), (dim=2|id=188|"S=1/2,Site,n=2"), (dim=3|id=992|"Link,l=2"))
[3] ((dim=3|id=992|"Link,l=2"), (dim=2|id=520|"S=1/2,Site,n=3"), (dim=3|id=41|"Link,l=3"))
[4] ((dim=3|id=41|"Link,l=3"), (dim=2|id=649|"S=1/2,Site,n=4"), (dim=3|id=261|"Link,l=4"))
[5] ((dim=3|id=261|"Link,l=4"), (dim=2|id=850|"S=1/2,Site,n=5"), (dim=2|id=776|"Link,l=5"))
[6] ((dim=2|id=776|"Link,l=5"), (dim=2|id=751|"S=1/2,Site,n=6"))


In [84]:
ψ4_dag = prime(linkinds, dag(ψ4))

MPS
[1] ((dim=2|id=113|"S=1/2,Site,n=1"), (dim=3|id=423|"Link,l=1")')
[2] ((dim=3|id=423|"Link,l=1")', (dim=2|id=188|"S=1/2,Site,n=2"), (dim=3|id=992|"Link,l=2")')
[3] ((dim=3|id=992|"Link,l=2")', (dim=2|id=520|"S=1/2,Site,n=3"), (dim=3|id=41|"Link,l=3")')
[4] ((dim=3|id=41|"Link,l=3")', (dim=2|id=649|"S=1/2,Site,n=4"), (dim=3|id=261|"Link,l=4")')
[5] ((dim=3|id=261|"Link,l=4")', (dim=2|id=850|"S=1/2,Site,n=5"), (dim=2|id=776|"Link,l=5")')
[6] ((dim=2|id=776|"Link,l=5")', (dim=2|id=751|"S=1/2,Site,n=6"))


In [85]:
L = ψ4[1] * ψ4_dag[1] * ψ4[2] * ψ4_dag[2] * ψ4[3] * ψ4_dag[3] * ψ4[4] * ψ4_dag[4] * ψ4[5] * ψ4_dag[5]

ITensor ord=2 (dim=2|id=776|"Link,l=5") (dim=2|id=776|"Link,l=5")'
NDTensors.Dense{Float64, Vector{Float64}}

In [86]:
out = L * ψ4[6] * ψ4[6]'

ITensor ord=2 (dim=2|id=751|"S=1/2,Site,n=6") (dim=2|id=751|"S=1/2,Site,n=6")'
NDTensors.Dense{Float64, Vector{Float64}}

In [87]:
@show out

out = ITensor ord=2
Dim 1: (dim=2|id=751|"S=1/2,Site,n=6")
Dim 2: (dim=2|id=751|"S=1/2,Site,n=6")'
NDTensors.Dense{Float64, Vector{Float64}}
 2×2
  0.5871178549425845   -0.0966833328705275
 -0.09668333287052748   0.4128821450574157


ITensor ord=2 (dim=2|id=751|"S=1/2,Site,n=6") (dim=2|id=751|"S=1/2,Site,n=6")'
NDTensors.Dense{Float64, Vector{Float64}}

---------

# Approach Trial

We will do a 10 site MPS

In [2]:
s = siteinds("S=1/2", 10)
ψ = randomMPS(s; linkdims=5)

MPS
[1] ((dim=2|id=698|"S=1/2,Site,n=1"), (dim=5|id=921|"Link,l=1"))
[2] ((dim=5|id=921|"Link,l=1"), (dim=2|id=624|"S=1/2,Site,n=2"), (dim=5|id=19|"Link,l=2"))
[3] ((dim=5|id=19|"Link,l=2"), (dim=2|id=697|"S=1/2,Site,n=3"), (dim=5|id=679|"Link,l=3"))
[4] ((dim=5|id=679|"Link,l=3"), (dim=2|id=603|"S=1/2,Site,n=4"), (dim=5|id=91|"Link,l=4"))
[5] ((dim=5|id=91|"Link,l=4"), (dim=2|id=438|"S=1/2,Site,n=5"), (dim=5|id=182|"Link,l=5"))
[6] ((dim=5|id=182|"Link,l=5"), (dim=2|id=566|"S=1/2,Site,n=6"), (dim=5|id=246|"Link,l=6"))
[7] ((dim=5|id=246|"Link,l=6"), (dim=2|id=247|"S=1/2,Site,n=7"), (dim=5|id=561|"Link,l=7"))
[8] ((dim=5|id=561|"Link,l=7"), (dim=2|id=995|"S=1/2,Site,n=8"), (dim=4|id=347|"Link,l=8"))
[9] ((dim=4|id=347|"Link,l=8"), (dim=2|id=494|"S=1/2,Site,n=9"), (dim=2|id=944|"Link,l=9"))
[10] ((dim=2|id=944|"Link,l=9"), (dim=2|id=187|"S=1/2,Site,n=10"))


Now let's measure site 1 to be "large" (max value)

In [6]:
large = Vector([1.0; 0])
projector = large * large'
projector_as_operator = op(projector, s[1])
@show projector_as_operator

projector_as_operator = ITensor ord=2
Dim 1: (dim=2|id=698|"S=1/2,Site,n=1")'
Dim 2: (dim=2|id=698|"S=1/2,Site,n=1")
NDTensors.Dense{Float64, Vector{Float64}}
 2×2
 1.0  0.0
 0.0  0.0


ITensor ord=2 (dim=2|id=698|"S=1/2,Site,n=1")' (dim=2|id=698|"S=1/2,Site,n=1")
NDTensors.Dense{Float64, Vector{Float64}}

Now we apply the single site operator at the first site to make the measurement, giving us a new MPS

In [10]:
site1_before_measure = deepcopy(ψ[1])
site1_after_measure = site1_before_measure * projector_as_operator
noprime!(site1_after_measure)

ITensor ord=2 (dim=5|id=921|"Link,l=1") (dim=2|id=698|"S=1/2,Site,n=1")
NDTensors.Dense{Float64, Vector{Float64}}

In [11]:
@show site1_after_measure

site1_after_measure = ITensor ord=2
Dim 1: (dim=5|id=921|"Link,l=1")
Dim 2: (dim=2|id=698|"S=1/2,Site,n=1")
NDTensors.Dense{Float64, Vector{Float64}}
 5×2
  0.06392531521821398   0.0
  0.44740667062379563   0.0
 -0.021142188400833956  0.0
  0.3166858817039392    0.0
 -0.4578842317775713    0.0


ITensor ord=2 (dim=5|id=921|"Link,l=1") (dim=2|id=698|"S=1/2,Site,n=1")
NDTensors.Dense{Float64, Vector{Float64}}

In [13]:
ψ[1] = site1_after_measure

ITensor ord=2 (dim=5|id=921|"Link,l=1") (dim=2|id=698|"S=1/2,Site,n=1")
NDTensors.Dense{Float64, Vector{Float64}}

In [16]:
normalize!(ψ)

MPS
[1] ((dim=5|id=921|"Link,l=1"), (dim=2|id=698|"S=1/2,Site,n=1"))
[2] ((dim=5|id=921|"Link,l=1"), (dim=2|id=624|"S=1/2,Site,n=2"), (dim=5|id=19|"Link,l=2"))
[3] ((dim=5|id=19|"Link,l=2"), (dim=2|id=697|"S=1/2,Site,n=3"), (dim=5|id=679|"Link,l=3"))
[4] ((dim=5|id=679|"Link,l=3"), (dim=2|id=603|"S=1/2,Site,n=4"), (dim=5|id=91|"Link,l=4"))
[5] ((dim=5|id=91|"Link,l=4"), (dim=2|id=438|"S=1/2,Site,n=5"), (dim=5|id=182|"Link,l=5"))
[6] ((dim=5|id=182|"Link,l=5"), (dim=2|id=566|"S=1/2,Site,n=6"), (dim=5|id=246|"Link,l=6"))
[7] ((dim=5|id=246|"Link,l=6"), (dim=2|id=247|"S=1/2,Site,n=7"), (dim=5|id=561|"Link,l=7"))
[8] ((dim=5|id=561|"Link,l=7"), (dim=2|id=995|"S=1/2,Site,n=8"), (dim=4|id=347|"Link,l=8"))
[9] ((dim=4|id=347|"Link,l=8"), (dim=2|id=494|"S=1/2,Site,n=9"), (dim=2|id=944|"Link,l=9"))
[10] ((dim=2|id=944|"Link,l=9"), (dim=2|id=187|"S=1/2,Site,n=10"))


Now that we have measured site 1, we would like to know the probability of site 2 being up or down, given we have measured site 1 to be large.

In [17]:
orthogonalize!(ψ, 2)

MPS
[1] ((dim=2|id=698|"S=1/2,Site,n=1"), (dim=2|id=186|"Link,l=1"))
[2] ((dim=2|id=624|"S=1/2,Site,n=2"), (dim=5|id=19|"Link,l=2"), (dim=2|id=186|"Link,l=1"))
[3] ((dim=5|id=19|"Link,l=2"), (dim=2|id=697|"S=1/2,Site,n=3"), (dim=5|id=679|"Link,l=3"))
[4] ((dim=5|id=679|"Link,l=3"), (dim=2|id=603|"S=1/2,Site,n=4"), (dim=5|id=91|"Link,l=4"))
[5] ((dim=5|id=91|"Link,l=4"), (dim=2|id=438|"S=1/2,Site,n=5"), (dim=5|id=182|"Link,l=5"))
[6] ((dim=5|id=182|"Link,l=5"), (dim=2|id=566|"S=1/2,Site,n=6"), (dim=5|id=246|"Link,l=6"))
[7] ((dim=5|id=246|"Link,l=6"), (dim=2|id=247|"S=1/2,Site,n=7"), (dim=5|id=561|"Link,l=7"))
[8] ((dim=5|id=561|"Link,l=7"), (dim=2|id=995|"S=1/2,Site,n=8"), (dim=4|id=347|"Link,l=8"))
[9] ((dim=4|id=347|"Link,l=8"), (dim=2|id=494|"S=1/2,Site,n=9"), (dim=2|id=944|"Link,l=9"))
[10] ((dim=2|id=944|"Link,l=9"), (dim=2|id=187|"S=1/2,Site,n=10"))


In [18]:
rho = prime(ψ[2], s[2]) * dag(ψ[2])

ITensor ord=2 (dim=2|id=624|"S=1/2,Site,n=2")' (dim=2|id=624|"S=1/2,Site,n=2")
NDTensors.Dense{Float64, Vector{Float64}}

In [19]:
@show rho

rho = ITensor ord=2
Dim 1: (dim=2|id=624|"S=1/2,Site,n=2")'
Dim 2: (dim=2|id=624|"S=1/2,Site,n=2")
NDTensors.Dense{Float64, Vector{Float64}}
 2×2
 0.419677230223639   0.1042551891926163
 0.1042551891926163  0.5803227697763612


ITensor ord=2 (dim=2|id=624|"S=1/2,Site,n=2")' (dim=2|id=624|"S=1/2,Site,n=2")
NDTensors.Dense{Float64, Vector{Float64}}

Let's get the trace and verify it is Hermitian

In [20]:
tr(rho)

1.0000000000000002

In [21]:
isequal(rho, dag(rho))

true

Now that we have the density matrix $\rho$, let's diagonalize to get the probabilities (eigenvalues) as well as the eigenstates:

In [47]:
rho_M = matrix(rho)

2×2 Matrix{Float64}:
 0.419677  0.104255
 0.104255  0.580323

In [50]:
F = eigen(rho_M)

Eigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}
values:
2-element Vector{Float64}:
 0.36839114080680396
 0.6316088591931962
vectors:
2×2 Matrix{Float64}:
 -0.897305  0.44141
  0.44141   0.897305

In [51]:
F.values

2-element Vector{Float64}:
 0.36839114080680396
 0.6316088591931962

In [53]:
sum(F.values)

1.0000000000000002

So the eigenvals sum to 1, which is what we expect since they reprent probabilities of the corresponding eigenstates.

So we have the following:
$$\rho = \lambda_1 \ket{v_1} \bra{v_1} + \lambda_2 \ket{v_2} \bra{v_2}$$
Where $\lambda_1 = 0.3683$, $\lambda_2 = 0.6316$

In [57]:
F.vectors[:, 1] *F.vectors[:, 1]' * F.values[1] + F.vectors[:, 2] * F.vectors[:, 2]' * F.values[2]

2×2 Matrix{Float64}:
 0.419677  0.104255
 0.104255  0.580323

In [58]:
F.vectors[:,1]

2-element Vector{Float64}:
 -0.8973054963693176
  0.4414100657953015

# Simulate an outcome
We now generate a random number and determine an outcome.

In [70]:
using Random

In [85]:
p1, p2 = F.values;

In [87]:
p1

0.36839114080680396

In [88]:
p2

0.6316088591931962

In [97]:
r = rand()
if r <= p1
    print("First eigenstate")
    outcome = F.vectors[:, 1]
else
    print("Second eigenstate")
    outcome = F.vectors[:,2]
end

First eigenstate

2-element Vector{Float64}:
 -0.8973054963693176
  0.4414100657953015

Here, we obtain the first eigenstate as the outcome from our random number generator. Thus, we collpase our quantum state onto the first eigenvector:

In [98]:
outcome

2-element Vector{Float64}:
 -0.8973054963693176
  0.4414100657953015

In [105]:
measurement = F.vectors[:,1]

2-element Vector{Float64}:
 -0.8973054963693176
  0.4414100657953015

Let's convert this back to a time-series value by getting an angle:

In [116]:
measurement

2-element Vector{Float64}:
 -0.8973054963693176
  0.4414100657953015

In [126]:
θ = acos(measurement[1]) 
(2 * θ)/π

1.7089568495750118

Now we construct a projector to project our MPS onto the subspace in which we make this measurement:

In [107]:
measurement_projector = measurement * measurement'
measurement_projector_as_operator = op(measurement_projector, s[2])

ITensor ord=2 (dim=2|id=624|"S=1/2,Site,n=2")' (dim=2|id=624|"S=1/2,Site,n=2")
NDTensors.Dense{Float64, Vector{Float64}}

In [114]:
site_2_before_measurment = deepcopy(ψ[2])
site_2_after_measurement = site_2_before_measurment * measurement_projector_as_operator
noprime!(site_2_after_measurement)

ITensor ord=3 (dim=5|id=19|"Link,l=2") (dim=2|id=186|"Link,l=1") (dim=2|id=624|"S=1/2,Site,n=2")
NDTensors.Dense{Float64, Vector{Float64}}

Update MPS with measured site 2

In [128]:
ψ[2] = site_2_after_measurement

ITensor ord=3 (dim=5|id=19|"Link,l=2") (dim=2|id=186|"Link,l=1") (dim=2|id=624|"S=1/2,Site,n=2")
NDTensors.Dense{Float64, Vector{Float64}}

In [130]:
normalize!(ψ)

MPS
[1] ((dim=2|id=698|"S=1/2,Site,n=1"), (dim=2|id=186|"Link,l=1"))
[2] ((dim=5|id=19|"Link,l=2"), (dim=2|id=186|"Link,l=1"), (dim=2|id=624|"S=1/2,Site,n=2"))
[3] ((dim=5|id=19|"Link,l=2"), (dim=2|id=697|"S=1/2,Site,n=3"), (dim=5|id=679|"Link,l=3"))
[4] ((dim=5|id=679|"Link,l=3"), (dim=2|id=603|"S=1/2,Site,n=4"), (dim=5|id=91|"Link,l=4"))
[5] ((dim=5|id=91|"Link,l=4"), (dim=2|id=438|"S=1/2,Site,n=5"), (dim=5|id=182|"Link,l=5"))
[6] ((dim=5|id=182|"Link,l=5"), (dim=2|id=566|"S=1/2,Site,n=6"), (dim=5|id=246|"Link,l=6"))
[7] ((dim=5|id=246|"Link,l=6"), (dim=2|id=247|"S=1/2,Site,n=7"), (dim=5|id=561|"Link,l=7"))
[8] ((dim=5|id=561|"Link,l=7"), (dim=2|id=995|"S=1/2,Site,n=8"), (dim=4|id=347|"Link,l=8"))
[9] ((dim=4|id=347|"Link,l=8"), (dim=2|id=494|"S=1/2,Site,n=9"), (dim=2|id=944|"Link,l=9"))
[10] ((dim=2|id=944|"Link,l=9"), (dim=2|id=187|"S=1/2,Site,n=10"))


In [131]:
norm(ψ)

0.9999999999999999

# Site 3

We will try one more site

(1) Get the reduced density matrix

In [132]:
ψ3 = orthogonalize(ψ, 3)

MPS
[1] ((dim=2|id=698|"S=1/2,Site,n=1"), (dim=2|id=186|"Link,l=1"))
[2] ((dim=2|id=186|"Link,l=1"), (dim=2|id=624|"S=1/2,Site,n=2"), (dim=4|id=819|"Link,l=2"))
[3] ((dim=2|id=697|"S=1/2,Site,n=3"), (dim=5|id=679|"Link,l=3"), (dim=4|id=819|"Link,l=2"))
[4] ((dim=5|id=679|"Link,l=3"), (dim=2|id=603|"S=1/2,Site,n=4"), (dim=5|id=91|"Link,l=4"))
[5] ((dim=5|id=91|"Link,l=4"), (dim=2|id=438|"S=1/2,Site,n=5"), (dim=5|id=182|"Link,l=5"))
[6] ((dim=5|id=182|"Link,l=5"), (dim=2|id=566|"S=1/2,Site,n=6"), (dim=5|id=246|"Link,l=6"))
[7] ((dim=5|id=246|"Link,l=6"), (dim=2|id=247|"S=1/2,Site,n=7"), (dim=5|id=561|"Link,l=7"))
[8] ((dim=5|id=561|"Link,l=7"), (dim=2|id=995|"S=1/2,Site,n=8"), (dim=4|id=347|"Link,l=8"))
[9] ((dim=4|id=347|"Link,l=8"), (dim=2|id=494|"S=1/2,Site,n=9"), (dim=2|id=944|"Link,l=9"))
[10] ((dim=2|id=944|"Link,l=9"), (dim=2|id=187|"S=1/2,Site,n=10"))


In [133]:
ρ = prime(ψ3[3], s[3]) * dag(ψ3[3])

ITensor ord=2 (dim=2|id=697|"S=1/2,Site,n=3")' (dim=2|id=697|"S=1/2,Site,n=3")
NDTensors.Dense{Float64, Vector{Float64}}

In [135]:
M = matrix(ρ)

2×2 Matrix{Float64}:
 0.362018  0.288146
 0.288146  0.637982

In [136]:
tr(M)

1.0000000000000002

In [138]:
F = eigen(M);

In [145]:
λ1, λ2 = F.values

2-element Vector{Float64}:
 0.18052037359804832
 0.8194796264019519

In [146]:
λ1

0.18052037359804832

In [147]:
v1 = F.vectors[:, 1]
v2 = F.vectors[:, 2]

2-element Vector{Float64}:
 0.5329646904863672
 0.8461374821474172

In [152]:
p1 = λ1

0.18052037359804832

In [153]:
r = rand()
if r <= p1
    print("First eigenstate")
    outcome = F.vectors[:, 1]
else
    print("Second eigenstate")
    outcome = F.vectors[:, 2]
end

Second eigenstate

2-element Vector{Float64}:
 0.5329646904863672
 0.8461374821474172

So now we measure the following value at site 3

In [156]:
θ = acos(outcome[1])
x = (2 * θ)/π 

0.6421557028748794

Now, construct the projector